### PR TITLE
Add reference tests for simple solver kernels

### PR DESCRIPTION
**Pull request by upsj**
To simplify #733, I pulled out the reference solver kernels which I added to check for detailed behavior of the kernels.

They check for correct default behavior, handling of stopped columns and division by zero.

### CODE REVIEW

**Review by thoasm — APPROVED** [review]
LGTM!

File: reference/test/solver/bicg_kernels.cpp [hunk 1]
```diff
@@ -117,6 +163,134 @@ class Bicg : public ::testing::Test {
...
+    this->small_rho->fill(1);
+    this->small_prev_rho->fill(0);
+
+    gko::kernels::reference::bicg::step_1(
+        this->exec, this->small_p.get(), this->small_z.get(),
+        this->small_p2.get(), this->small_z2.get(), this->small_rho.get(),
+        this->small_prev_rho.get(), &this->small_stop);
+
+    GKO_ASSERT_MTX_NEAR(this->small_p, l({{-2.0, -2.0}, {-2.0, -2.0}}), 0);
+    GKO_ASSERT_MTX_NEAR(this->small_p2, l({{-2.0, -2.0}, {-2.0, -2.0}}), 0);
```
Comment (MarcelKoch) [inline reference/test/solver/bicg_kernels.cpp:241]: Would it make sense to use the following change to clarify the expected behavior?

```suggestion
    GKO_ASSERT_MTX_NEAR(this->small_p, this->small_z, 0);
    GKO_ASSERT_MTX_NEAR(this->small_p2, this->small_z2, 0);
```

File: reference/test/solver/bicgstab_kernels.cpp [hunk 2]
```diff
@@ -108,6 +163,249 @@ class Bicgstab : public ::testing::Test {
...
+    this->small_alpha->at(0) = 1;
+    this->small_alpha->at(1) = -2;
+
+    gko::kernels::reference::bicgstab::step_3(
+        this->exec, this->small_x.get(), this->small_r.get(),
+        this->small_s.get(), this->small_t.get(), this->small_y.get(),
+        this->small_z.get(), this->small_alpha.get(), this->small_beta.get(),
+        this->small_gamma.get(), this->small_omega.get(), &this->small_stop);
+
+    GKO_ASSERT_MTX_NEAR(this->small_x, l({{9.0, -3.0}, {9.0, -3.0}}), 0);
```
Comment (MarcelKoch) [inline reference/test/solver/bicgstab_kernels.cpp:385]: I think the check for `this->small_r` is missing. It should be
```suggestion
    GKO_ASSERT_MTX_NEAR(this->small_x, l({{9.0, -3.0}, {9.0, -3.0}}), 0);
    GKO_ASSERT_MTX_NEAR(this->small_r, this->small_s, 0);
```

**Review by MarcelKoch — COMMENTED** [review]
lgtm, perhaps the changes I mentioned could be adopted, although I don't think that is necessary.

**Review by MarcelKoch — APPROVED** [review]
lgtm

**Review by pratikvn — APPROVED** [review]
LGTM!

**Pull request merged**